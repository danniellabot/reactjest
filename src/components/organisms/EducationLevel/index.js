// a dropdown for selecting Education Level
// use Autocomplete from material-ui
// Typography from material-ui for the label of the dropdown menu
// Value is from context
// onChange, call dispatch with the new value
// options = high school, college, graduate school, other

import * as React from "react";
import TextField from "@mui/material/TextField";
import Autocomplete from "@mui/material/Autocomplete";
import Typography from "@mui/material/Typography";
import { StateContext, DispatchContext } from "../../../context/Provider";

const options = ["None", "High School", "College", "Graduate School", "Other"];

export default function EducationLevel() {
  const [inputValue, setInputValue] = React.useState("");
  const [show, setShow] = React.useState(false);
  const state = React.useContext(StateContext);
  const dispatch = React.useContext(DispatchContext);
  const [value, setValue] = React.useState(state.educationLevel);

  const handleChange = (event, newValue) => {
    setValue(newValue);
    dispatch({ type: "UPDATE_EDUCATION_LEVEL", payload: newValue });
  };

  const handleClickOpen = () => {
    setShow(true);
  };

  return (
    <div>
      <div>{`value: ${value !== null ? `'${value}'` : "null"}`}</div>
      <div>{`inputValue: '${inputValue}'`}</div>
      <div>{`inputValue: '${state.educationLevel}'`}</div>
      <br />
      <Typography variant="h6"
        sx={{ mt: 3, mb: 1, fontWeight: "bold" }}
        onClick={handleClickOpen}
        data-testid="education-level-label"
      >{state.educationLevel}</Typography>
      {show && (
      <Autocomplete
        value={value}
        onChange={handleChange}
        inputValue={inputValue}
        onInputChange={(event, newInputValue) => {
          setInputValue(newInputValue);
        }}
        data-testid="education-level-select"
        id="controllable-states-demo"
        options={options}
        sx={{ width: 300 }}
        renderInput={(params) => <TextField {...params} label="Controllable" />}
      />
      )}
    </div>
  );
}
