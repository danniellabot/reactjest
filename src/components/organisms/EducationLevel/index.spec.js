// unit test for Age component
// Author: Jansen Lefever
// Date: 2020-01-15
// License: MIT

import React, { useReducer, useState } from "react";
import renderer from "react-test-renderer";
import {
  render,
  fireEvent,
  renderHook,
  screen,
  within,
} from "@testing-library/react";
import {
  DispatchContext,
  StateContext,
  reducer,
} from "../../../context/Provider";
import EducationLevel from "./index";

describe("EducationLevel", () => {
  // mock state to pass to the component
  const mockState = {
    educationLevel: "None",
  };

  // mock dispatch to pass to the component
  const mockDispatch = jest.fn();

  // wrapper to pass to the component
  const wrapper = ({ children }) => (
    <StateContext.Provider value={mockState}>
      <DispatchContext.Provider value={mockDispatch}>
        {children}
      </DispatchContext.Provider>
    </StateContext.Provider>
  );

  const handleClickOpen = jest.fn();
  const handleClose = jest.fn();

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <StateContext.Provider value={mockState}>
          <DispatchContext.Provider value={mockDispatch}>
            <EducationLevel />
          </DispatchContext.Provider>
        </StateContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  /**
   * Test that onClick state.educationLevel, useState is called to set to open
   */
  it("should set open to true when clicked", () => {
    const { getByTestId } = render(<EducationLevel />, { wrapper });
    const educationLevel = getByTestId("education-level-label");
    fireEvent.click(educationLevel);
    expect(handleClickOpen).toHaveBeenCalled();
  });

  /**
   * Test that the component renders but does not show the dropdown
   */
  it("renders", () => {
    const tree = renderer.create(<EducationLevel />, { wrapper }).toJSON();
    expect(tree).toMatchSnapshot();
  });

  /**
   * Test that the component renders and shows the dropdown
   * when the user clicks on the label
   * and that the dropdown has the correct options
   */
  it("renders and shows the dropdown", () => {
    const { getByText } = render(<EducationLevel />, { wrapper });
    const label = getByText("Education Level");
    fireEvent.click(label);
    const dropdown = screen.getByTestId("education-level-select");
    const { getAllByRole } = within(dropdown);
    const options = getAllByRole("option");
    expect(options).toHaveLength(5);
    expect(options[0]).toHaveTextContent("None");
  });

  /**
   * Test that the component changes the state when the user selects a different option in the dropdown
   */
  test("EducationLevel changes state when user selects a different option", () => {
    const { getByTestId } = render(
      <DispatchContext.Provider value={mockDispatch}>
        <StateContext.Provider value={mockState}>
          <EducationLevel />
        </StateContext.Provider>
      </DispatchContext.Provider>
    );
    const select = getByTestId("education-level-select");
    const input = within(select).getByRole("combobox");
    select.focus();
    fireEvent.change(input, { target: { value: "High School" } });
    fireEvent.keyDown(select, { key: "ArrowDown" });
    fireEvent.keyDown(select, { key: "Enter" });
    expect(input).toHaveValue("High School");
    expect(mockDispatch).toHaveBeenCalledWith({
      type: "UPDATE_EDUCATION_LEVEL",
      payload: "High School",
    });
  });
});
