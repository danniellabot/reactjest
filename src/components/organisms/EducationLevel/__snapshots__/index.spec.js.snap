// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EducationLevel renders correctly 1`] = `
<div>
  <div>
    value: 'None'
  </div>
  <div>
    inputValue: ''
  </div>
  <div>
    inputValue: 'None'
  </div>
  <br />
  <h6
    className="MuiTypography-root MuiTypography-h6 css-u6xmfb-MuiTypography-root"
    onClick={[Function]}
  >
    Education Level
  </h6>
</div>
`;

exports[`EducationLevel renders with the correct initial state 1`] = `
<div>
  <div>
    value: 'None'
  </div>
  <div>
    inputValue: ''
  </div>
  <div>
    inputValue: 'None'
  </div>
  <br />
  <h6
    className="MuiTypography-root MuiTypography-h6 css-u6xmfb-MuiTypography-root"
    onClick={[Function]}
  >
    Education Level
  </h6>
</div>
`;
